README
NAME

    Alien::Libarchive - Build and make available libarchive

VERSION

    version 0.22

SYNOPSIS

    Build.PL

     use Alien::Libarchive;
     use Module::Build;
     
     my $alien = Alien::Libarchive->new;
     my $build = Module::Build->new(
       ...
       extra_compiler_flags => $alien->cflags,
       extra_linker_flags   => $alien->libs,
       ...
     );
     
     $build->create_build_script;

    Makefile.PL

     use Alien::Libarchive;
     use ExtUtils::MakeMaker;
     
     my $alien = Alien::Libarchive;
     WriteMakefile(
       ...
       CFLAGS => $alien->cflags,
       LIBS   => $alien->libs,
     );

    FFI::Raw

     use Alien::Libarchive;
     use FFI::Raw;
     
     my($dll) = Alien::Libarchive->new->dlls;
     FFI::Raw->new($dll, 'archive_read_new', FFI::Raw::ptr);

    FFI::Sweet

     use Alien::Libarchive;
     use FFI::Sweet;
     
     ffi_lib( Alien::Libarchive->new->dlls );
     attach_function 'archive_read_new', [], _ptr;

DESCRIPTION

    This distribution installs libarchive so that it can be used by other
    Perl distributions. If already installed for your operating system, and
    it can be found, this distribution will use the libarchive that comes
    with your operating system, otherwise it will download it from the
    internet, build and install it.

    If you set the environment variable ALIEN_LIBARCHIVE to 'share', this
    distribution will ignore any system libarchive found, and build from
    source instead. This may be desirable if your operating system comes
    with a very old version of libarchive and an upgrade path for the
    system libarchive is not possible.

 Requirements

  operating system install

    The development headers and libraries for libarchive

    Debian

      On Debian you can install these with this command:

       % sudo apt-get install libarchive-dev

    Cygwin

      On Cygwin, make sure that this package is installed

       libarchive-devel

    FreeBSD

      libarchive comes with FreeBSD as of version 5.3.

  from source install

    A C compiler and any prerequisites for building libarchive.

    Debian

      On Debian build-essential should be good enough:

       % sudo apt-get install build-essential

    Cygwin

      On Cygwin, I couldn't get libarchive to build without making a minor
      tweak to one of the include files. On Cygwin this module will patch
      libarchive before it attempts to build if it is version 3.1.2.

METHODS

 cflags

    Returns the C compiler flags necessary to build against libarchive.

    Returns flags as a list in list context and combined into a string in
    scalar context.

 libs

    Returns the library flags necessary to build against libarchive.

    Returns flags as a list in list context and combined into a string in
    scalar context.

 dlls

    Returns a list of dynamic libraries (usually a list of just one
    library) that make up libarchive. This can be used for FFI::Raw.

    Returns just the first dynamic library found in scalar context.

 version

    Returns the libarchive version.

 install_type

    Returns the install type, one of either system or share.

 pkg_config_dir

    Returns a path that contains the libarchive.pc file which can be used
    by pkg-config for linking against libarchive.

 pkg_config_name

    Returns the name by which pkg-config knows libarchive (should always be
    libarchive).

CAVEATS

    Debian Linux and FreeBSD (9.0) have been tested the most in development
    of this distribution.

    Patches to improve portability and platform support would be eagerly
    appreciated.

    If you reinstall this distribution, you may need to reinstall any
    distributions that depend on it as well.

SEE ALSO

    Alien::Libarchive::Installer

    Archive::Libarchive::XS

    Archive::Libarchive::FFI

    Archive::Libarchive::Any

    Archive::Ar::Libarchive

    Archive::Peek::Libarchive

    Archive::Extract::Libarchive

    YAPC::NA 2014 Foreign Function Interface (FFI) : Never Need to Write XS
    Again
    <https://docs.google.com/presentation/d/1NY3ROAiSAC5yk1LoeBCM5JfAmSeTdo
    pgYFnJO6mUtXI/edit?usp=sharing>

      Slides on a talk about FFI and Alien

AUTHOR

    Graham Ollis <plicease@cpan.org>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2014 by Graham Ollis.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Alien-Libarchive

0.22      2014-10-08 14:34:37 -0400
  - add version method
  - fixed weird recursive deps on MSWin32

0.21      2014-06-19 07:41:16 -0400
  - promoted to production

0.20_04   2014-06-18 16:02:34 -0400
  - explicitly declare Alien::bz2::Installer as a configure prereq

0.20_03   2014-06-18 07:55:33 -0400
  - use Alien::LZO::Installer to find / install libzo2 if possible

0.20_02   2014-06-17 03:49:45 -0400
  - mock up a libbz2.la on MSWin32 so that bzip2 + libarchive dll works

0.20_01   2014-06-16 18:21:55 -0400
  - use Alien::bz2::Installer if possible to build libbz2 if it is not provided by OS

0.20      2014-06-10 20:32:27 -0400
  - mark prereqs as dynamic

0.19      2014-06-10 20:17:45 -0400
  - dynamic libs handle upgrades more gracefully
  - promote to production

0.18_06   2014-06-10 10:39:36 -0400
  - Archive::Libarchive::XS dev compat

0.18_05   2014-06-10 10:19:42 -0400
  - require Alien::Libarchive::Installer 0.03

0.18_04   2014-06-10 09:19:32 -0400
  - cflags reference bug fix

0.18_03   2014-06-10 09:06:55 -0400
  - fix prereqs

0.18_02   2014-06-10 08:36:20 -0400
  - libs and cflags now return a list in list context and a scalar in scalar context
    (used to always return a scalar)

0.18_01   2014-06-10 08:19:36 -0400
  - completely new implementation based on Alien::Libarchive::Installer instead of Alien::Base

0.18      2014-02-03 16:40:43 -0500
  - Fix ExtUtils::CChecker calling error

0.17      2014-02-03 16:37:39 -0500
  - remove -fPIC / -fno-common detection and instead rely on $Config{cccdlflags}
  - use $Config{optimize} for CFLAGS passed to libarchive configure

0.16      2014-02-02 04:30:52 -0500
  - build with -fPIC if possible

0.15      2014-01-29 11:30:49 -0500
  - fix prereqs

0.14      2014-01-29 09:37:52 -0500
  - build with -fno-common if possible

0.12      2014-01-02 10:59:11 -0500
  - require Alien::Libarchive::MSWin32 0.04 on MSWin32

0.11      2013-12-24 09:26:53 -0500
  - require Alien::Libarchive::MSWin32 0.02 on MSWin32

0.10      2013-12-19 03:04:41 -0500
  - fix for beter detection of libarchive sans pkg-config as a system install
    (runtime didn't work on non-FreeBSD systems)

0.09      2013-12-18 09:06:08 -0500
  - verify Alien deps flags can at least compile a simple c program
  - better detection of libarchive sans pkg-config as a system install

0.08      2013-12-12 17:07:56 -0500
  - documentation
  - try to use Alien::bz2 if available
  - include bin directory in PATH for cygwin install_type = share 
    so dll can be found
  - basic compile check to be sure it works

0.07      2013-12-05 18:44:24 -0500
  - prefer non-system lib when that is what you installed
  - patch 3.1.2 on cygwin so that it will build
  - Strawberry Perl support via Alien::Libarchive::MSWin32

0.06      2013-11-29 10:02:22 -0500
  - support for perl 5.12, 5.10, 5.8

0.05      2013-11-26 16:57:45 -0500
  - more diagnostics

0.04      2013-11-26 12:05:20 -0500
  - use Alien::OpenSSL and Alien::LibXML if available
  - First CPAN release

0.03      2013-11-20 12:46:58 -0500
  - support for FreeBSD native build

0.02      2013-11-15 10:55:29 -0500
  - rm deps on Alien::OpenSSL, Alien::LibXML
  - rm dep on Path::Class

0.01      2013-11-14 17:07:12 -0500
  - initial version


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
use strict;
use warnings;

use Module::Build 0.3601;
use lib qw{inc}; use My::ModuleBuild;

my %module_build_args = (
  "build_requires" => {
    "Alien::Libarchive::Installer" => "0.06",
    "Alien::bz2::Installer" => "0.03",
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Alien::Libarchive::Installer" => "0.06",
    "Alien::bz2::Installer" => "0.03",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Build and make available libarchive",
  "dist_author" => [
    "Graham Ollis <plicease\@cpan.org>"
  ],
  "dist_name" => "Alien-Libarchive",
  "dist_version" => "0.22",
  "license" => "perl",
  "module_name" => "Alien::Libarchive",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Alien::Libarchive::Installer" => "0.06",
    "File::ShareDir" => 0,
    "perl" => "5.006"
  },
  "script_files" => [],
  "share_dir" => {
    "dist" => "share"
  },
  "test_requires" => {
    "Test::More" => "0.94"
  }
);


my %fallback_build_requires = (
  "Alien::Libarchive::Installer" => "0.06",
  "Alien::bz2::Installer" => "0.03",
  "Module::Build" => "0.3601",
  "Test::More" => "0.94"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = My::ModuleBuild->new(%module_build_args);

$build->create_build_script;


inc/My/ModuleBuild.pm
package My::ModuleBuild;

use strict;
use warnings;
use base qw( Module::Build );
use Config;
use Alien::Libarchive::Installer;
use Alien::bz2::Installer;
use Alien::LZO::Installer;
use File::Spec;
use FindBin ();

my $type = eval { require FFI::Raw } ? 'both' : 'compile';

# Note: for historical / hysterical reasons, the install type is one of:
# 1. system, use the system libarchive
# 2. share, build your own libarchive, both static and shared
#    the static version will be used for XS modules and the shared one
#    will be used for FFI modules.

sub _list ($)
{
  ref($_[0]) eq 'ARRAY' ? $_[0] : [$_[0]];
}

sub _catfile {
  my $path = File::Spec->catfile(@_);
  $path =~ s{\\}{/}g if $^O eq 'MSWin32';
  $path;
}

sub _catdir {
  my $path = File::Spec->catdir(@_);
  $path =~ s{\\}{/}g if $^O eq 'MSWin32';
  $path;
}

sub new
{
  my($class, %args) = @_;

  my $system;

  unless(($ENV{ALIEN_LIBARCHIVE} || 'system') eq 'share')
  {
    $system = eval {
      Alien::Libarchive::Installer->system_install(
        type  => $type,
        alien => 0,
      )
    };
  }

  unless(defined $system)
  {
    foreach my $class (map { "Alien::$_\::Installer" } qw( Libarchive bz2 LZO ))
    {
      my $prereqs = $class->build_requires;  
      while(my($mod,$ver) = each %$prereqs)
      {
        $args{build_requires}->{$mod} = $ver;
      }
    }
  }

  my $self = $class->SUPER::new(%args);

  $self->config_data( name => 'libarchive' );
  $self->config_data( already_built => 0 );
  $self->config_data( msvc => $^O eq 'MSWin32' && $Config{cc} =~ /cl(\.exe)?$/i ? 1 : 0 );
  
  $self->add_to_cleanup( '_alien', '_alien_libarchive', '_alien_bz2', '_alien_lzo', 'share/libarchive019' );
  
  if(defined $system)
  {
    print "Found libarchive " . $system->version . " from system\n";
    print "You can set ALIEN_LIBARCHIVE=share to force building from source\n";
    $self->config_data( install_type => 'system' );
    $self->config_data( cflags       => _list $system->cflags );
    $self->config_data( libs         => _list $system->libs );
    $self->config_data( version      => $system->version );
  }
  else
  {
    print "Did not find working libarchive, will download and install from the Internet\n";
    $self->config_data( install_type => 'share' );
  }
  
  $self;
}

sub ACTION_build
{
  my $self = shift;
  
  if($self->config_data('install_type') eq 'share')
  {
    unless($self->config_data('already_built'))
    {
      my $prefix = _catdir($FindBin::Bin, 'share', 'libarchive019' );

      local $ENV{CONFIG_SITE} = _catfile($FindBin::Bin, '_alien', 'config.site');
      do {
        my $lib = _catdir($prefix, 'lib');
        my $inc = _catdir($prefix, 'include');
        my $dll = _catdir($prefix, 'dll');
      
        mkdir(_catdir($FindBin::Bin, '_alien'));
        open my $fh, '>', $ENV{CONFIG_SITE};
        print $fh "#!/bin/sh\n",
                  "if [ -n \"\$CPPFLAGS\" ] ; then\n",
                  "  CPPFLAGS=\"\$CPPFLAGS -I$inc\"\n",
                  "else\n",
                  "  CPPFLAGS=\"-I$inc\"\n",
                  "fi;\n",
                  "if [ -n \"\$LDFLAGS\" ] ; then\n",
                  "  LDFLAGS=\"\$LDFLAGS -L$lib -L$dll\"\n",
                  "else\n",
                  "  LDFLAGS=\"-L$lib -L$dll\"\n",
                  "fi;\n";
        close $fh;
      };

      if(eval { require Alien::LZO::Installer; })
      {
        my $build = eval { Alien::LZO::Installer->system_install };
        unless($build)
        {
          my $build_dir = _catdir($FindBin::Bin, '_alien_lzo');
          mkdir $build_dir unless -d $build_dir;
          $build = eval { Alien::LZO::Installer->build_install($prefix, dir => $build_dir) };
        }
      }

      if(eval { require Alien::bz2::Installer; })
      {
        my $build = eval { Alien::bz2::Installer->system_install };
        unless($build)
        {
          my $build_dir = _catdir($FindBin::Bin, '_alien_bz2');
          mkdir $build_dir unless -d $build_dir;
          $build = eval { Alien::bz2::Installer->build_install($prefix, dir => $build_dir) };
        }
        
        if(defined $build)
        {
          if($^O eq 'MSWin32')
          {
            my $dir = _catdir($prefix, 'dll');
            mkdir $dir;
            my $la = _catfile($dir, 'libbz2.la');
            open my $fh, '>', $la;
            binmode $fh;
            print $fh _bz2_la_file($prefix);
            close $fh;
          }
        }
      }
    
      my $build_dir = _catdir($FindBin::Bin, '_alien_libarchive');
      mkdir $build_dir unless -d $build_dir;
      mkdir $prefix unless -d $prefix;
      my $build = Alien::Libarchive::Installer->build_install( $prefix, dir => $build_dir );
      $self->config_data( cflags => [grep !/^-I/, @{ _list $build->cflags }] );
      $self->config_data( libs =>   [grep !/^-L/, @{ _list $build->libs }] );
      if($self->config_data('msvc'))
      {
        $self->config_data( libs =>   [grep !/^(\/|-)libpath/i, @{ _list $build->libs }] );
      }
      $self->config_data( version => $build->version );

      printf "cflags: %s\n", join ' ', @{ $self->config_data('cflags') };
      printf "libs:   %s\n", join ' ', @{ $self->config_data('libs') };
      printf "msvc:   %d\n", $self->config_data('msvc');
      
      do {
        opendir my $dh, _catdir($prefix, 'dll');
        my @list = grep { ! -l _catfile($prefix, 'dll', $_) }
                   grep { /\.so/ || /\.(dll|dylib)$/ }
                   grep !/^(libbz2|bzip2.dll|liblzo2)/,
                   grep !/^\./,
                   sort
                   readdir $dh;
        closedir $dh;
        print "dlls:\n";
        print "  - $_\n" for @list;
        $self->config_data( dlls => \@list );
      };
      
      $self->config_data( already_built => 1 );
    }
  }
  
  $self->SUPER::ACTION_build(@_);
}

my $bz2_la_template;
sub _bz2_la_file
{
  my($prefix) = @_;
  unless(defined $bz2_la_template)
  {
    $bz2_la_template = do { local $/; <DATA> };
  }
  my $bz2_la = $bz2_la_template;
  $bz2_la =~ s{##PREFIX##}{$prefix}g;
  $bz2_la;
}

1;

__DATA__
# libbz2.la - a libtool library file
# Generated by libtool (GNU libtool) 2.4.2 Debian-2.4.2-1.1
#
# Please DO NOT delete this file!
# It is necessary for linking the library.

# The name that we can dlopen(3).
dlname='../bin/bzip2.dll'

# Names of this library.
library_names='../dll/libbz2.dll.a'

# The name of the static archive.
old_library='libbz2.a'

# Linker flags that can not go in dependency_libs.
inherited_linker_flags=''

# Libraries that this one depends upon.
dependency_libs=''

# Names of additional weak libraries provided by this library
weak_library_names=''

# Version information for libbz2.
current=14
age=1
revision=2

# Is this an already installed library?
installed=yes

# Should we warn about portability when linking against -modules?
shouldnotlink=no

# Files to dlopen/dlpreopen
dlopen=''
dlpreopen=''

# Directory that this library needs to be installed in:
libdir='##PREFIX##/lib'


lib/Alien/Libarchive/Unix.pm
package Alien::Libarchive::Unix;

use strict;
use warnings;
use base qw( Alien::Libarchive );

# ABSTRACT: Build and make available libarchive on Unix
our $VERSION = '0.22'; # VERSION


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Alien::Libarchive::Unix - Build and make available libarchive on Unix

=head1 VERSION

version 0.22

=head1 DESCRIPTION

This module is deprecated and kept only for backwards
compatibility.  You should use L<Alien::Libarchive>
instead.

=head1 SEE ALSO

=over 4

=item L<Alien::Libarchive>

=item L<Archive::Libarchive::XS>

=item L<Archive::Libarchive::FFI>

=item L<Archive::Libarchive::Any>

=item L<Archive::Ar::Libarchive>

=item L<Archive::Peek::Libarchive>

=item L<Archive::Extract::Libarchive>

=back

=head1 AUTHOR

Graham Ollis <plicease@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Alien/Libarchive/MSWin32.pm
package Alien::Libarchive::MSWin32;

use strict;
use warnings;
use base qw( Alien::Libarchive );

# ABSTRACT: Build and make available libarchive on Windows
our $VERSION = '0.22'; # VERSION


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Alien::Libarchive::MSWin32 - Build and make available libarchive on Windows

=head1 VERSION

version 0.22

=head1 DESCRIPTION

This module is deprecated and kept only for backwards
compatibility.  You should use L<Alien::Libarchive>
instead.

=head1 SEE ALSO

=over 4

=item L<Alien::Libarchive>

=item L<Archive::Libarchive::XS>

=item L<Archive::Libarchive::FFI>

=item L<Archive::Libarchive::Any>

=item L<Archive::Ar::Libarchive>

=item L<Archive::Peek::Libarchive>

=item L<Archive::Extract::Libarchive>

=back

=head1 AUTHOR

Graham Ollis <plicease@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Alien/Libarchive.pm
package Alien::Libarchive;

use strict;
use warnings;
use File::ShareDir ();
use File::Spec;
use Alien::Libarchive::ConfigData;
use constant _share_dir => File::ShareDir::dist_dir('Alien-Libarchive');
use constant _alien_libarchive019 => 1;

# ABSTRACT: Build and make available libarchive
our $VERSION = '0.22'; # VERSION

my $cf = 'Alien::Libarchive::ConfigData';

sub _catfile {
  my $path = File::Spec->catfile(@_);
  $path =~ s{\\}{/}g if $^O eq 'MSWin32';
  $path;
}

sub _catdir {
  my $path = File::Spec->catdir(@_);
  $path =~ s{\\}{/}g if $^O eq 'MSWin32';
  $path;
}


sub new
{
  my($class) = @_;
  bless {}, $class;
}


sub cflags
{
  my($class) = @_;
  my @cflags = @{ $cf->config("cflags") };
  unshift @cflags, '-I' . _catdir(_share_dir, 'libarchive019', 'include' )
    if $class->install_type eq 'share';
  wantarray ? @cflags : "@cflags";
}


sub libs
{
  my($class) = @_;
  my @libs = @{ $cf->config("libs") };
  if($class->install_type eq 'share')
  {
    if($cf->config('msvc'))
    {
      unshift @libs, '/libpath:' . _catdir(_share_dir, 'libarchive019', 'lib');
      @libs = map { s{^.*(\\|/)}{} if m/archive_static\.lib$/; $_ } @libs;
    }
    else
    {
      unshift @libs, '-L' . _catdir(_share_dir, 'libarchive019', 'lib');
    }
  }
  wantarray ? @libs : "@libs";
}


sub dlls
{
  my($class) = @_;
  my @list;
  if($class->install_type eq 'system')
  {
    require Alien::Libarchive::Installer;
    @list = Alien::Libarchive::Installer->system_install( alien => 0, test => 'ffi' )->dlls;
  }
  else
  {
    @list = map { _catfile(_share_dir, 'libarchive019', 'dll', $_) }
            @{ $cf->config("dlls") };
  }
  wantarray ? @list : $list[0];
}


sub version
{
  $cf->config("version");
}


sub install_type
{
  $cf->config("install_type");
}


sub pkg_config_dir
{
  _catdir(_share_dir, 'libarchive019', 'lib', 'pkgconfig');
}


sub pkg_config_name
{
  'libarchive';
}

# extract the macros from the header files, this is a private function
# because it may not be portable.  Used by the Archive::Libarchive::XS
# build process (and maybe Archive::Libarchive::FFI) to automatically
# generate constants
# UPDATE: this maybe should use C::Scan or C::Scan::Constants
sub _macro_list
{
  require Config;
  require File::Temp;
  require File::Spec;

  my $alien = Alien::Libarchive->new;
  my $cc = "$Config::Config{ccname} $Config::Config{ccflags} " . $alien->cflags;

  my $fn = File::Spec->catfile(File::Temp::tempdir( CLEANUP => 1 ), "test.c");

  do {
    open my $fh, '>', $fn;
    print $fh "#include <archive.h>\n";
    print $fh "#include <archive_entry.h>\n";
    close $fh;
  };

  my @list;
  my $cmd = "$cc -E -dM $fn";
  foreach my $line (`$cmd`)
  {
    if($line =~ /^#define ((AE|ARCHIVE)_\S+)/)
    {
      push @list, $1;
    }
  }
  sort @list;
}


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Alien::Libarchive - Build and make available libarchive

=head1 VERSION

version 0.22

=head1 SYNOPSIS

Build.PL

 use Alien::Libarchive;
 use Module::Build;
 
 my $alien = Alien::Libarchive->new;
 my $build = Module::Build->new(
   ...
   extra_compiler_flags => $alien->cflags,
   extra_linker_flags   => $alien->libs,
   ...
 );
 
 $build->create_build_script;

Makefile.PL

 use Alien::Libarchive;
 use ExtUtils::MakeMaker;
 
 my $alien = Alien::Libarchive;
 WriteMakefile(
   ...
   CFLAGS => $alien->cflags,
   LIBS   => $alien->libs,
 );

FFI::Raw

 use Alien::Libarchive;
 use FFI::Raw;
 
 my($dll) = Alien::Libarchive->new->dlls;
 FFI::Raw->new($dll, 'archive_read_new', FFI::Raw::ptr);

FFI::Sweet

 use Alien::Libarchive;
 use FFI::Sweet;
 
 ffi_lib( Alien::Libarchive->new->dlls );
 attach_function 'archive_read_new', [], _ptr;

=head1 DESCRIPTION

This distribution installs libarchive so that it can be used by other Perl
distributions.  If already installed for your operating system, and it can
be found, this distribution will use the libarchive that comes with your
operating system, otherwise it will download it from the internet, build
and install it.

If you set the environment variable ALIEN_LIBARCHIVE to 'share', this
distribution will ignore any system libarchive found, and build from
source instead.  This may be desirable if your operating system comes
with a very old version of libarchive and an upgrade path for the 
system libarchive is not possible.

=head2 Requirements

=head3 operating system install

The development headers and libraries for libarchive

=over 4

=item Debian

On Debian you can install these with this command:

 % sudo apt-get install libarchive-dev

=item Cygwin

On Cygwin, make sure that this package is installed

 libarchive-devel

=item FreeBSD

libarchive comes with FreeBSD as of version 5.3.

=back

=head3 from source install

A C compiler and any prerequisites for building libarchive.

=over 4

=item Debian

On Debian build-essential should be good enough:

 % sudo apt-get install build-essential

=item Cygwin

On Cygwin, I couldn't get libarchive to build without making a
minor tweak to one of the include files.  On Cygwin this module
will patch libarchive before it attempts to build if it is
version 3.1.2.

=back

=head1 METHODS

=head2 cflags

Returns the C compiler flags necessary to build against libarchive.

Returns flags as a list in list context and combined into a string in
scalar context.

=head2 libs

Returns the library flags necessary to build against libarchive.

Returns flags as a list in list context and combined into a string in
scalar context.

=head2 dlls

Returns a list of dynamic libraries (usually a list of just one library)
that make up libarchive.  This can be used for L<FFI::Raw>.

Returns just the first dynamic library found in scalar context.

=head2 version

Returns the libarchive version.

=head2 install_type

Returns the install type, one of either C<system> or C<share>.

=head2 pkg_config_dir

Returns a path that contains the libarchive.pc file which can be used
by pkg-config for linking against libarchive.

=head2 pkg_config_name

Returns the name by which pkg-config knows libarchive (should always
be libarchive).

=head1 CAVEATS

Debian Linux and FreeBSD (9.0) have been tested the most
in development of this distribution.

Patches to improve portability and platform support would be eagerly
appreciated.

If you reinstall this distribution, you may need to reinstall any
distributions that depend on it as well.

=head1 SEE ALSO

=over 4

=item L<Alien::Libarchive::Installer>

=item L<Archive::Libarchive::XS>

=item L<Archive::Libarchive::FFI>

=item L<Archive::Libarchive::Any>

=item L<Archive::Ar::Libarchive>

=item L<Archive::Peek::Libarchive>

=item L<Archive::Extract::Libarchive>

=item L<YAPC::NA 2014 Foreign Function Interface (FFI) : Never Need to Write XS Again|https://docs.google.com/presentation/d/1NY3ROAiSAC5yk1LoeBCM5JfAmSeTdopgYFnJO6mUtXI/edit?usp=sharing>

Slides on a talk about FFI and Alien

=back

=head1 AUTHOR

Graham Ollis <plicease@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
